packages/moonstone/VideoPlayer/Video.js

```diff
@@ -0,0 +1,297 @@
+
+		/**
+		 * The video source to be preloaded. Expects a `<source>` node.
+		 *
+		 * @type {String|Node}
+		 * @public
+		 */
+		preloadSource:  PropTypes.node,
+
+		/**
+		 * Called with a reference to the active [Media]{@link ui/Media.Media} component
+		 *
+		 * @type {Function}
+		 * @private
+		 */
+		setMedia: PropTypes.func,
+
+		/**
+		 * The video source to be played.
+		 *
+		 * Any children `<source>` elements will be sent directly to the `mediaComponent` as video
+		 * sources.
+		 *
+		 * See: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/source
+		 *
+		 * @type {Node}
+		 * @public
+		 */
+		source: PropTypes.node
+	}
+
+	static defaultProps = {
+		mediaComponent: 'video'
+	}
+
+	constructor (props) {
+		super(props);
+		this.isPlayerMounted = false;
+
+		this.loaded = {
+			video: false,
+			preload: false
+		};
+	}
+
+	componentDidMount () {
+		this.isPlayerMounted = true;
+		this.setMedia();
+	}
+
+	componentDidUpdate (prevProps) {
+		const {source, preloadSource} = this.props;
+		const {source: prevSource, preloadSource: prevPreloadSource} = prevProps;
+
+		const key = getKeyFromSource(source);
+		const prevKey = getKeyFromSource(prevSource);
+		const preloadKey = getKeyFromSource(preloadSource);
+		const prevPreloadKey = getKeyFromSource(prevPreloadSource);
+
+		// if there's source and it has changed and it's not the prior preloaded source
+		if (source && key !== prevKey) {
+			// if it wasn't the preload key
+			if (key !== prevPreloadKey) {
+				// flag it as unloaded
+				this.loaded.video = false;
+			} else {
+				// if the preload video is switched to the current, try to play it now because it
+				// won't fire an onLoadStart events since it has already loaded.
+				this.autoPlay();
+			}
+		}
+
+		if (preloadSource && preloadKey !== prevPreloadKey) {
+			// flag it as unloaded (if its not the same as source) and load it
+			this.loaded.preload = preloadKey === key;
+
+			// In the case that the previous source equalled the previous preload (causing the
+			// preload video node to not be created) and then the preload source was changed, we
+			// need to guard against accessing the preloadVideo node.
+			if (this.preloadVideo) {
+				this.preloadVideo.load();
+			}
+		}
+
+		if (this.props.setMedia !== prevProps.setMedia) {
+			this.clearMedia(prevProps);
+			this.setMedia();
+		}
+	}
+
+	componentWillUnmount () {
+		this.clearMedia();
+	}
+
+	canPlay () {
+		return this.isPlayerMounted && this.loaded.video && (
+			// video can play if there isn't a preloaded source or there is and it has loaded
+			!this.props.preloadSource || this.loaded.preload
+		);
+	}
+
+	clearMedia ({setMedia} = this.props) {
+		if (setMedia) {
+			setMedia(null);
+		}
+	}
+
+	setMedia ({setMedia} = this.props) {
+		if (setMedia) {
+			setMedia(this.video);
+		}
+	}
+
+	autoPlay () {
+		if (this.props.noAutoPlay || !this.canPlay()) return;
+
+		this.video.play();
+
+		forward('onLoadStart', this.loadStartEvent, this.props);
```

Comment:
It seems like `loadstart` event was fired from the preload source would be still forwarded but with the payload from the other event. Any specific reason we want to do this?

Reply:
I don't think that will happen. If the preload video fires `onLoadStart` first, `canPlay` will prevent this line from being reached because `this.loaded.video` will be `false`. If the video fires `onLoadStart` first, its event will be cached in `loadStartEvent` and will be forwarded when the preload video fires `onLoadStart`.

Do you read it the same way?